File: app/src/main/java/com/mandarin/bcu/androidutil/io/asynchs/DownloadApk.java
package com.mandarin.bcu.androidutil.io.asynchs;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import androidx.core.content.FileProvider;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.mandarin.bcu.R;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

public class DownloadApk extends AsyncTask<Void,Integer,Void> {
    private final WeakReference<Activity> weakReference;
    private final String ver;
    private final String url;
    private final String path;
    private final String realpath;
    private File output = null;

    public DownloadApk(Activity context,String ver,String url,String path,String realpath) {
        this.weakReference = new WeakReference<>(context);
        this.ver = ver;
        this.url = url;
        this.path = path;
        this.realpath =realpath;
    }

    @Override
    protected void onPreExecute() {
        Activity activity = weakReference.get();

        if(activity == null) return;

        TextView state = activity.findViewById(R.id.apkstate);

        if(state != null)
            state.setText(R.string.down_wait);
    }

    @Override
    protected Void doInBackground(Void... voids) {
        try {
            URL realurl = new URL(url);
            HttpURLConnection c = (HttpURLConnection)realurl.openConnection();
            c.setRequestMethod("GET");
            c.connect();

            Long size = (long)c.getContentLength();

            output = new File(realpath);
            File pathes = new File(path);

            if(!pathes.exists()) {
                pathes.mkdirs();
            }

            if(!output.exists()) {
                output.createNewFile();
            }

            FileOutputStream fos = new FileOutputStream(output);
            InputStream is = c.getInputStream();

            byte [] buffer = new byte[1024];
            int len1;
            long total = (long)0;

            while((len1 = is.read(buffer))!= -1) {
                total += len1;
                int progress = (int)(total*100/size);
                publishProgress(progress);
                fos.write(buffer,0,len1);
            }

            c.disconnect();
            fos.close();
            is.close();
        } catch (MalformedURLException e) {
            e.printStackTrace();
            output = null;
        } catch (IOException e) {
            e.printStackTrace();
            output = null;
        }

        return null;
    }

    @Override
    protected void onProgressUpdate(Integer... values) {
        Activity activity = weakReference.get();

        if(activity == null) return;

        ProgressBar prog = activity.findViewById(R.id.apkprog);
        TextView state = activity.findViewById(R.id.apkstate);
        String name = activity.getString(R.string.down_state_doing)+"BCU_Android_"+ver+".apk";

        if(state != null)
            state.setText(name);

        if(prog != null) {
            if (prog.isIndeterminate()) {
                prog.setIndeterminate(false);
            }

            prog.setProgress(values[0]);
        }
    }

    @Override
    protected void onPostExecute(Void result) {
        Activity activity = weakReference.get();

        if(activity == null) return;

        if(output == null) {
            Button retry = activity.findViewById(R.id.apkretry);

            if(retry != null)
                retry.setVisibility(View.VISIBLE);
        } else {
            File install = new File(realpath);
            Uri apkuri = FileProvider.getUriForFile(activity, "com.mandarin.bcu.provider",install);
            Intent intent = new Intent(Intent.ACTION_INSTALL_PACKAGE);
            intent.setData(apkuri);
            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            activity.startActivity(intent);
            activity.finish();
        }
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/io/asynchs/DownloadApk.kt
package com.mandarin.bcu.androidutil.io.asynchs

import android.app.Activity
import android.content.ContentValues
import android.content.Intent
import android.content.pm.PackageInstaller
import android.content.pm.PackageManager
import android.net.Uri
import android.os.AsyncTask
import android.os.Environment
import android.provider.MediaStore
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.FileProvider
import androidx.documentfile.provider.DocumentFile
import com.mandarin.bcu.R
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.lang.ref.WeakReference
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URI
import java.net.URL

class DownloadApk(context: Activity, private val ver: String, private val url: String, private val path: String, private val realpath: String) : AsyncTask<Void?, Int?, Void?>() {
    private val weakReference: WeakReference<Activity> = WeakReference(context)
    private var output: File? = null
    override fun onPreExecute() {
        val activity = weakReference.get() ?: return
        val state = activity.findViewById<TextView>(R.id.apkstate)
        state?.setText(R.string.down_wait)
    }

    override fun doInBackground(vararg voids: Void?): Void? {
        try {
            val realurl = URL(url)
            val c = realurl.openConnection() as HttpURLConnection
            c.requestMethod = "GET"
            c.connect()
            val size = c.contentLength.toLong()
            output = File(realpath)

            val pathes = File(path)

            if (!pathes.exists()) {
                pathes.mkdirs()
            }
            if (!output!!.exists()) {
                output!!.createNewFile()
            }

            val fos = FileOutputStream(output)
            val `is` = c.inputStream
            val buffer = ByteArray(1024)
            var len1: Int
            var total = 0.toLong()
            while (`is`.read(buffer).also { len1 = it } != -1) {
                total += len1.toLong()
                val progress = (total * 100 / size).toInt()
                publishProgress(progress)
                fos.write(buffer, 0, len1)
            }
            c.disconnect()
            fos.close()
            `is`.close()
        } catch (e: MalformedURLException) {
            e.printStackTrace()
            output = null
        } catch (e: IOException) {
            e.printStackTrace()
            output = null
        }
        return null
    }

    override fun onProgressUpdate(vararg values: Int?) {
        val activity = weakReference.get() ?: return
        val prog = activity.findViewById<ProgressBar>(R.id.apkprog)
        val state = activity.findViewById<TextView>(R.id.apkstate)
        val name = activity.getString(R.string.down_state_doing) + "BCU_Android_" + ver + ".apk"
        if (state != null) state.text = name
        if (prog != null) {
            if (prog.isIndeterminate) {
                prog.isIndeterminate = false
            }
            prog.progress = values[0] ?: 0
        }
    }

    override fun onPostExecute(result: Void?) {
        val activity = weakReference.get() ?: return
        if (output == null) {
            val retry = activity.findViewById<Button>(R.id.apkretry)
            if (retry != null) retry.visibility = View.VISIBLE
        } else {
            val install = File(realpath)
            val apkuri = FileProvider.getUriForFile(activity, "com.mandarin.bcu.provider", install)

            val intent = Intent(Intent.ACTION_VIEW).setDataAndType(apkuri, "application/vnd.android.package-archive")
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)

            activity.startActivity(intent)
            activity.finish()
        }
    }
}